sctp/src/association/mod.rs

```diff
@@ -498,36 +498,68 @@ impl Association {
         mut awake_write_loop_ch: mpsc::Receiver<()>,
     ) {
         log::debug!("[{}] write_loop entered", name);
-        let mut done = false;
-        while !done {
+        let done = Arc::new(AtomicBool::new(false));
+        let name = Arc::new(name);
+
+        let limit = {
+            #[cfg(test)]
+            {
+                1
+            }
+            #[cfg(not(test))]
+            {
+                8
+            }
+        };
+
+        let sem = Arc::new(Semaphore::new(limit));
+        while !done.load(Ordering::Relaxed) {
             //log::debug!("[{}] gather_outbound begin", name);
-            let (raw_packets, mut ok) = {
+            let (packets, continue_loop) = {
                 let mut ai = association_internal.lock().await;
```

Comment:
I feel like there has to be a way for us to remove this lock completely... we just need a bounded channel or lock-free queue to add and remove items. No need to lock entire association

Reply:
I agree, and it doesn't even need to be bounded, the amount of packets the can be in this channel would be naturally limited by the amount of bytes allowed to be in flight by the normal SCTP mechanisms.

I just didn't want to change to architecture too much.